qiskit/backends/local/unitary_simulator_py.py

```diff
@@ -128,19 +130,60 @@ def _add_unitary_single(self, gate, qubit):
             is q_{n-1} ... otimes q_1 otimes q_0.
         number_of_qubits is the number of qubits in the system.
         """
-        unitary_add = enlarge_single_opt(gate, qubit, self._number_of_qubits)
-        self._unitary_state = np.dot(unitary_add, self._unitary_state)
 
-    def _add_unitary_two(self, gate, q_0, q_1):
+        # Convert to complex rank-2 tensor
+        gate_tensor = np.array(gate, dtype=complex)
+
+        # Compute einsum index string
```

Comment:
Non-blocking: Einstein's summation convention seems really powerful! Could you extract these index calculations into a couple of functions and provide short explanations. A super-brief introduction of the `einsum` function would be great for the coders, like me, that are not experts in the matter. But I'm not blocking this, I understand this is a very specific code.